lib/fs.js

```diff
@@ -1836,22 +1836,36 @@ fs.FileWriteStream = fs.WriteStream; // support the legacy name
 
 
 WriteStream.prototype.open = function() {
+  const self = this;
+
   fs.open(this.path, this.flags, this.mode, function(er, fd) {
     if (er) {
-      this.destroy();
-      this.emit('error', er);
+      self.destroy();
+      self.emit('error', er);
       return;
     }
 
-    this.fd = fd;
-    this.emit('open', fd);
-  }.bind(this));
+    self.fd = fd;
+    self.emit('open', fd);
+  });
 };
 
 
+function write(fd, buffer, position, callback) {
+  function wrapper(err, written) {
+    // Retain a reference to buffer so that it can't be GC'ed too soon.
+    callback(err, written || 0, buffer);
+  }
+
+  var req = new FSReqWrap();
+  req.oncomplete = wrapper;
+  binding.writeBuffer(fd, buffer, 0, buffer.length, position, req);
+}
+
+
 WriteStream.prototype._write = function(data, encoding, cb) {
   if (!(data instanceof Buffer))
-    return this.emit('error', new Error('Invalid data'));
+    return this.emit('error', new TypeError('Invalid data'));
```

Comment:
I'm pretty sure this isn't necessary anyways, as Writable guarantees data will be a Buffer. Perhaps, if the C++ throws on not-a-buffer, this could be removed entirely? (As to throw on evil `._write` calls)